File: Niveau 3/Niveau 3.11/rallonges.py
#-------------------------------------------------------------------------------
# Name:        Rallonges audio
# Purpose:     France-IOI niveau 3.11
#
# Author:      Sébastien
#
# Created:     28/10/2022
#-------------------------------------------------------------------------------

def main():
    n = int(input())
    longueur = 0
    femelles = []
    males = []
    for _ in range(n):
        c1, c2, l = map(int, input().split())
        if c1 != c2:
            longueur += l
        elif c1 == 0:
            femelles.append(l)
        else:
            males.append(l)

    femelles.sort(reverse=True)
    males.sort(reverse=True)

    if len(males) > 0:
        n = min(len(femelles), len(males)-1)
        longueur += sum(femelles[:n]) + sum(males[:n+1])
        print(longueur)
    else:
        print(-1)

if __name__ == '__main__':
    main()

"""
5
1 0 1000
1 1 500
0 0 2000
1 1 800
0 0 3000
---
7
0 0 500
1 1 1000
0 0 600
1 1 2000
0 0 300
0 0 400
0 0 200
"""

File: Niveau 3/Niveau 3.11/rallonges.hs
--------------------------------------------------------------------------------
-- Name:        Rallonges audio
-- Purpose:     France-IOI niveau 3.11
--
-- Author:      Sébastien
--
-- Created:     29/10/2022
--------------------------------------------------------------------------------

import Control.Monad
import Data.List

main :: IO ()
main = do
    _ <- readLn :: IO Int
    lignes <- fmap lines getContents
    let (longueur, femelles, males) = foldr lireLigne (0, [], []) lignes
        femelles' = sort femelles
        males' = sort males
        f = length femelles'
        m = length males'
        n = min f (m - 1)
    if n > -1 then
        print $ longueur + sum (drop (m - 1 - n) males') + sum (drop (f - n) femelles')
    else print (-1)
    
    
lireLigne :: String -> (Int, [Int], [Int]) -> (Int, [Int], [Int])
lireLigne ligne (longueur, femelles, males) =
    let (c1:c2:l:_) = map read (words ligne) :: [Int] in
    if c1 /= c2 then (longueur+l, femelles, males)
    else if c1 == 0 then (longueur, l:femelles, males)
    else (longueur, femelles, l:males)
    
  